src/ot/fallback.rs

```diff
@@ -0,0 +1,521 @@
+    pos.x_offset = 0;
+    pos.y_offset = 0;
+
+    // We don't position LEFT and RIGHT marks.
+
+    // X positioning
+    match combining_class {
+        Class::DoubleBelow |
+        Class::DoubleAbove if direction.is_horizontal() => {
+            pos.x_offset += base_extents.x_bearing
+                + if direction.is_forward() { base_extents.width } else { 0 }
+                - mark_extents.width / 2 - mark_extents.x_bearing;
+        }
+
+        Class::AttachedBelowLeft |
+        Class::BelowLeft |
+        Class::AboveLeft => {
+            // Left align.
+            pos.x_offset += base_extents.x_bearing - mark_extents.x_bearing;
+        }
+
+        Class::AttachedAboveRight |
+        Class::BelowRight |
+        Class::AboveRight => {
+            // Right align.
+            pos.x_offset += base_extents.x_bearing + base_extents.width
+                - mark_extents.width - mark_extents.x_bearing;
+        }
+
+        Class::AttachedBelow |
+        Class::AttachedAbove |
+        Class::Below |
+        Class::Above |
+        _ => {
+            // Center align.
+            pos.x_offset += base_extents.x_bearing
+                + (base_extents.width - mark_extents.width) / 2
+                - mark_extents.x_bearing;
+        }
+    }
+
+    let is_attached = matches!(
+        combining_class,
+        Class::AttachedBelowLeft |
+        Class::AttachedBelow |
+        Class::AttachedAbove |
+        Class::AttachedAboveRight
+    );
+
+    // Y positioning.
+    match combining_class {
+        Class::DoubleBelow |
+        Class::BelowLeft |
+        Class::Below |
+        Class::BelowRight |
+        Class::AttachedBelowLeft |
+        Class::AttachedBelow => {
+            if !is_attached {
+                // Add gap.
+                base_extents.height -= y_gap;
+            }
+
+            pos.y_offset = base_extents.y_bearing + base_extents.height
+                - mark_extents.y_bearing;
+
+            // Never shift up "below" marks.
+            if (y_gap > 0) == (pos.y_offset > 0) {
+                base_extents.height -= pos.y_offset;
+                pos.y_offset = 0;
+            }
+
+            base_extents.height += mark_extents.height;
+        }
+
+        Class::DoubleAbove |
+        Class::AboveLeft |
+        Class::Above |
+        Class::AboveRight |
+        Class::AttachedAbove |
+        Class::AttachedAboveRight => {
+            if !is_attached {
+                // Add gap.
+                base_extents.y_bearing += y_gap;
+                base_extents.height -= y_gap;
+            }
+
+            pos.y_offset = base_extents.y_bearing
+                - (mark_extents.y_bearing + mark_extents.height);
+
+            // Don't shift down "above" marks too much.
+            if (y_gap > 0) != (pos.y_offset > 0) {
+                // NOTE(laurmaedje): In the original this was "unsigned int".
+                // Does it matter? I don't think so, I have to cast back anyway, otherwise.
```

Comment:
Looks like a minor hb bug. I think it should be reported just in case.

Reply:
Done.